test-data/unit/check-python38.test

```diff
@@ -181,3 +181,122 @@ def f(p1: bytes, p2: float, /, p_or_kw: int) -> None:
+    reveal_type(y3)  # E: Revealed type is 'builtins.int'
+
+    # https://www.python.org/dev/peps/pep-0572/#exceptional-cases
+    (y4 := 3)
+    reveal_type(y4)  # E: Revealed type is 'builtins.int'
+
+    y5 = (y6 := 3)
+    reveal_type(y5)  # E: Revealed type is 'builtins.int'
+    reveal_type(y6)  # E: Revealed type is 'builtins.int'
+
+    f(x=(y7 := 3))
+    reveal_type(y7)  # E: Revealed type is 'builtins.int'
+
+    reveal_type((lambda: (y8 := 3) and y8)())  # E: Revealed type is 'builtins.int'
+
+    y7 = 1.0  # E: Incompatible types in assignment (expression has type "float", variable has type "int")
+    if y7 := "x":  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+        pass
+
+    # Just make sure we don't crash on this sort of thing.
+    if (NT := NamedTuple("NT", [("x", int)])):  # E: "int" not callable
+        z2: NT  # E: Invalid type "NT"
+
+    if (Alias := int):
+        z3: Alias  # E: Invalid type "Alias"
+
+    return (y8 := 3) + y8
+
+reveal_type(c)  # E: Revealed type is 'builtins.int'
+
+[builtins fixtures/f_string.pyi]
+
+[case testWalrusNewSemanal]
+# flags: --new-semantic-analyzer
+from typing import NamedTuple
+
+if (a := 2):
+    reveal_type(a)  # E: Revealed type is 'builtins.int'
+
+while (b := "x"):
+    reveal_type(b)  # E: Revealed type is 'builtins.str'
+
+def f(x: int = (c := 4)) -> int:
+    if (a := 2):
+        reveal_type(a)  # E: Revealed type is 'builtins.int'
+
+    while (b := "x"):
+        reveal_type(b)  # E: Revealed type is 'builtins.str'
+
+    x = (y := 1) + (z := 2)
+    reveal_type(x)  # E: Revealed type is 'builtins.int'
+    reveal_type(y)  # E: Revealed type is 'builtins.int'
+    reveal_type(z)  # E: Revealed type is 'builtins.int'
+
+    l = [y2 := 1, y2 + 2, y2 + 3]
+    reveal_type(y2)  # E: Revealed type is 'builtins.int'
+    reveal_type(l)  # E: Revealed type is 'builtins.list[builtins.int*]'
+
+    filtered_data = [y3 for x in l if (y3 := a) is not None]
+    reveal_type(filtered_data)  # E: Revealed type is 'builtins.list[builtins.int*]'
+    reveal_type(y3)  # E: Revealed type is 'builtins.int'
+
+    # https://www.python.org/dev/peps/pep-0572/#exceptional-cases
+    (y4 := 3)
+    reveal_type(y4)  # E: Revealed type is 'builtins.int'
+
+    y5 = (y6 := 3)
+    reveal_type(y5)  # E: Revealed type is 'builtins.int'
+    reveal_type(y6)  # E: Revealed type is 'builtins.int'
+
+    f(x=(y7 := 3))
+    reveal_type(y7)  # E: Revealed type is 'builtins.int'
+
+    reveal_type((lambda: (y8 := 3) and y8)())  # E: Revealed type is 'builtins.int'
+
+    y7 = 1.0  # E: Incompatible types in assignment (expression has type "float", variable has type "int")
+    if y7 := "x":  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+        pass
+
+    # Just make sure we don't crash on this sort of thing.
+    if (NT := NamedTuple("NT", [("x", int)])):  # E: "int" not callable
```

Comment:
This is probably because we define special forms as `NamedTuple = 0` in the stub. If we think users are likely to actually try this, I can work on a better error message.